### ISSUE TITLE
Not performing a dry-run: kubermatic-installer mirror-images --dry-run

### ISSUE THREAD

**Issue by judge-red** [kubermatic/kubermatic#11699]
### What happened?

`kubermatic-installer mirror-images --dry-run`, specifically the flag at the end, does not do what it says.

### Expected behavior

```
$ kubermatic-installer mirror-images --help | grep dry-run
      --dry-run                 Only print the names of found images
```

### How to reproduce the issue?

```
# no target registry should be required for a dry-run
$ kubermatic-installer mirror-images --dry-run
ERRO[0000] ❌ Operation failed: no target registry was passed. 

# this seems to be used to read custom registry information, but I want the upstream urls printed
# so this should not be required either for a dry-run (and I would argue not otherwise either)
$ kubermatic-installer mirror-images --dry-run foo.registry.bar
ERRO[0000] ❌ Operation failed: please specify your KubermaticConfiguration via --config. 

# Now it's attempting to pull images instead of just printing them, that's definitely not what the help says
$ kubermatic-installer mirror-images --dry-run --config kubermatic.yaml foo.registry.bar
INFO[0000] Extracting addon manifests from image…        image="some.registry.internal/foo/bar/custom-addons:v2.21.4" temp-directory=/var/folders/m2/0x9pjrf10m3gd9h0v77xwgr40000gp/T/imageloader608481999
INFO[0000] Downloading image…                            image="some.registry.internal/foo/bar/custom-addons:v2.21.4"
INFO[0000] Executing command…                            command="docker pull some.registry.internal/foo/bar/custom-addons:v2.21.4" image="some.registry.internal/foo/bar/custom-addons:v2.21.4"
ERRO[0000] Command failed: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?  command="docker pull some.registry.internal/foo/bar/custom-addons:v2.21.4" image="some.registry.internal/foo/bar/custom-addons:v2.21.4"
ERRO[0000] ❌ Operation failed: failed to create local addons path: failed to download addons image: failed to download some.registry.internal/foo/bar/custom-addons:v2.21.4: failed to pull image some.registry.internal/foo/bar/custom-addons:v2.21.4: exit status 1. 
```

### How is your environment configured?

- KKP version: 2.21.4 EE
- Shared or separate master/seed clusters?: N/A

### Provide your KKP manifest here (if applicable)

N/A

### What cloud provider are you running on?

N/A

### What operating system are you running in your user cluster?

N/A but running `kubermatic-installer` on darwin amd64

**Comment by embik** [comment]
I don't think this is a bug. It needs to download the addons image to extract the actual list of images. So what you're seeing is not part of the mirroring but of the preparations to do so. That's intended behaviour.

If you don't want it to download any images, you can use `--addons-path` to pass an `addons` folder instead.

**Comment by embik** [comment]
Regarding the flags: dry runs are supposed to simulate the actual runs, and those need the respective flags and args. We probably need to improve the help message here, but the dry-run output prints both the origin and target registry so you can see where the image will be mirrored to.

**Comment by judge-red** [comment]
Hm, I see. `--addons-path` works for me, that's the behavior I expected, I guess. Just wished the addons were included in the release tarball and that this was the default action for `--dry-run`, but if you think this is how it should be done, I can work with it.

**Comment by embik** [comment]
I would love to improve the usability for `mirror-images` so the feedback here is very welcome.

To make sure I get this right, what do you consider the issue with fetching the addons image to extract it specifically? Is it the dependency on `docker`? Would you consider the default behaviour still fetching the image but with an embedded client that doesn't need `docker` to be an improvement? Or is the fact that it fetches the image the problem itself (apart from what we discussed in #11700 regarding the image source)?

**Comment by judge-red** [comment]
> To make sure I get this right, what do you consider the issue with fetching the addons image to extract it specifically? Is it the dependency on docker? 

Well, that's a good question.

Initially I thought it would go on to pull all images, i.e. I didn't know that it needs this image to source the rest of the images and that it will actually do a dry-run after that. So I think the output could be clearer, because again, it doesn't look like a dry-run and that could lead me to panic ctrl+c.

But yea, I'm on macOS and Docker Desktop (since the docker daemon will also be needed to pull images) comes with a licensing fee, so I don't use it. I think I could work around that with other tooling but it's rather cumbersome. I tried using Rancher Desktop and specifying `nerdctl` as the docker binary and then pulling worked - but running the container to extract info didn't for some reasons (didn't investigate as I didn't see this lead anywhere anymore and didn't even understand why it was trying to run the container at that time). I'm not sure what a good alternative to extract the info would be.

For the actual mirroring, I'd prefer if a tool was used that can do it directly, without pulling them to the local registry first. `skopeo` is a tool that could do that, but I guess more people have a proper Docker installation than skopeo, so yea. Anyway, this is how I do it in a docker-less root-less CI pipeline container (i.e. that's how we currently mirror some images):
```
skopeo copy --retry-times=3 --format=oci --src-creds="redacted:${DOCKER_HUB_PASSWORD}" --dest-creds="${CI_REGISTRY_USER}:${CI_REGISTRY_PASSWORD}" "docker://docker.io/${CI_JOB_NAME}:${IMG_TAG}" "docker://${CI_REGISTRY_IMAGE}/${CI_JOB_NAME}:${IMG_TAG}"
```

> Would you consider the default behaviour still fetching the image but with an embedded client that doesn't need docker to be an improvement?

That would be an improvement, but as per what I said above, extracting from the data from the image is the main pain point. But really, why not just include the addons in the release tarball (or extract and include the image during the release process)? `kubermatic-installer version` also doesn't work without the `charts` directory that is included.

Anyway, with the knowledge I now have, I can work with it - just have to add an extra step of cloning the repo and checking out the correct tag - so maybe just make it clearer how to use it or what it does.

Last but not least, it's fairly common to only mirror the Docker Hub registry, because of their rate limiting. For security reasons, it's best practice (and we certainly plan to implement that in future) to mirror and scan all images, but the first step for most is just avoiding Docker Hub's rate limit. To that end, it would be nice if I could tell the `kubermatic-installer` to only mirror certain registries.

**Comment by embik** [comment]
Great feedback, thanks so much for taking the time to write this all down. I'll take a look at how to improve the command based on that.

A few loosely collected answers/points:

- The `addons` folder isn't included in the release tarball because the image is already considered the release artefact. It would only be useful for this use case where extracting the image isn't an option. Maybe we need to reconsider this.
- I think mirroring a subset of the images based on the source registry makes sense. I'll look into it.
- I agree that `skopeo` is the better tool for this, but as you said it probably does not have a wide adoption among users. But perhaps using the libraries backing skopeo's copy functionality is an option. Something I'll investigate.
- I'll also check if we have an alternative to actually running the addons image to extract addon data.

**Comment by embik** [comment]
> Last but not least, it's fairly common to only mirror the Docker Hub registry, because of their rate limiting. For security reasons, it's best practice (and we certainly plan to implement that in future) to mirror and scan all images, but the first step for most is just avoiding Docker Hub's rate limit. To that end, it would be nice if I could tell the kubermatic-installer to only mirror certain registries.

While I see the use case here, this would need a fairly broad change in KKP as for example for addons, a singular override registry is expected/supported. It doesn't make much sense for the `mirror-images` command to support mirroring only for a specific registry when the other parts of KKP would not support such a mode of operation.

**Comment by judge-red** [comment]
> While I see the use case here, this would need a fairly broad change in KKP as for example for addons, a singular override registry is expected/supported.

Oh, I thought this had finally been addressed :/
Had to look it up (we had an RFE support ticket for this), but surely this applies to addons as well:
https://docs.kubermatic.com/kubermatic/v2.21/references/crds/#nodesettings

But yea, I agree it doesn't make sense for `mirror-images` to support this if KKP can't run with partial mirrors.

**Comment by embik** [comment]
> > While I see the use case here, this would need a fairly broad change in KKP as for example for addons, a singular override registry is expected/supported.
> 
> Oh, I thought this had finally been addressed :/ Had to look it up (we had an RFE support ticket for this), but surely this applies to addons as well: https://docs.kubermatic.com/kubermatic/v2.21/references/crds/#nodesettings
> 
> But yea, I agree it doesn't make sense for `mirror-images` to support this if KKP can't run with partial mirrors.

My bad - I'll be completely honest; I forgot about those settings and was only looking at `KubermaticConfiguration` options. You're right of course and thus, this should be possible. I'll look into a registry filter flag.

**Comment by judge-red** [comment]
> I agree that skopeo is the better tool for this, but as you said it probably does not have a wide adoption among users. But perhaps using the libraries backing skopeo's copy functionality is an option. Something I'll investigate.

Not sure copying what skopeo does is a good idea - unless 95% of it are really in a library. Maybe make it a flag to use skopeo for direct-mirroring instead of docker pull/push? So people who use skopeo and don't like the local pulling/pushing can opt for this. But not a flag of having too many flags either.

On the other hand, requiring skopeo to be available for `mirror-images` isn't that bad of a requirement imho. It's a small tool, easy to install. At least on macOS and Linux it's readily available, not sure about Windows.

**Comment by judge-red** [comment]
> My bad - I'll be completely honest, I forgot about those settings and was only looking at KubermaticConfiguration options.

No worries, I would have looked in the same place if I didn't have it linked in that ticket ;) Can't remember all the options of all the components :)

**Comment by embik** [comment]
@judge-red with #11705, we removed the dependency on `docker`, which includes the need to have it for downloading the addons image during dry-run. As such, with KKP 2.22 you won't need to pass the `addons` directory anymore.

By your judgement, would you say that resolves this issue and we can go ahead and close it?

**Comment by judge-red** [comment]
It's always a bit hard to judge such fixes before going hands-on with them. But I'd say yes - not promising I will open more issues in future, though :) Putting mirror-images into our CI pipeline is now definitely on our todo list ;)

Thanks for the work you put into improving the experience with this tool!

**Comment by embik** [comment]
> It's always a bit hard to judge such fixes before going hands-on with them. But I'd say yes - not promising I will open more issues in future, though :) Putting mirror-images into our CI pipeline is now definitely on our todo list ;)
> 
> Thanks for the work you put into improving the experience with this tool!

Makes total sense. I'll provisionally close this ticket as there are no further PRs planned, but please feel free to re-open it (or open new ones, as you said) if the iteration included in 2.22 is still missing feature for you.

Thanks for all the feedback on this tool, it helped a lot improving it!

**Comment by embik** [comment]
Sorry, I just realised I linked the wrong PR earlier on. #11717 removed the dependency on `docker`.
